File: core/backend/app/workspaces/system_core.py
"""
System-core workspace registration.

On backend startup (when ENABLE_SYSTEM_EVOLUTION=true), validates the
agent-space worktree is a git repo, finds/creates the owner's personal space,
and registers the workspace with:
  - workspace_type = system_core
  - protected = true
  - system_managed = true
  - registered_from = "auto"

The worktree path is derived from AGENT_SPACE_HOME: AGENT_SPACE_HOME/workspaces/<space_id>/agent-space

This module is ONLY invoked at startup — not through any API endpoint.
"""
from __future__ import annotations

import logging
import subprocess
from pathlib import Path

from sqlalchemy.orm import Session
from ulid import ULID

from ..config import settings, paths
from ..models import Space, SpaceMembership, User, Workspace

log = logging.getLogger(__name__)

# Well-known ID for the system-core workspace (stable across restarts)
SYSTEM_CORE_WORKSPACE_ID = "system-core-workspace"


def _new_id() -> str:
    return str(ULID())


def _validate_git_repo(path: Path) -> bool:
    """Return True if path is a valid git repository."""
    if not path.is_dir():
        return False
    try:
        result = subprocess.run(
            ["git", "rev-parse", "--git-dir"],
            cwd=path,
            capture_output=True,
            text=True,
            timeout=10,
        )
        return result.returncode == 0
    except (subprocess.TimeoutExpired, FileNotFoundError):
        return False


def _ensure_personal_space(db: Session, user: User) -> Space:
    """Find or create a personal space for the user."""
    # Look for existing personal space where user is owner/member
    membership = (
        db.query(SpaceMembership)
        .filter(
            SpaceMembership.user_id == user.id,
            SpaceMembership.role == "owner",
            SpaceMembership.status == "active",
        )
        .first()
    )
    if membership:
        space = db.query(Space).filter(Space.id == membership.space_id).first()
        if space and space.type == "personal":
            return space

    # Create a new personal space
    space = Space(
        id=_new_id(),
        name=f"{user.display_name}'s Personal Space",
        type="personal",
        created_by_user_id=user.id,
    )
    db.add(space)
    db.flush()
    db.add(SpaceMembership(
        id=_new_id(),
        space_id=space.id,
        user_id=user.id,
        role="owner",
        status="active",
    ))
    log.info("Created personal space '%s' for user %s", space.name, user.email)
    return space


def _call_deployer_init_worktree(repo_path: str) -> bool:
    """Ask deployer to clone the canonical repo into the worktree dir."""
    from ..deployment.client import DeployerClient
    from ..config import settings

    log.info("calling deployer init for worktree at %s", repo_path)
    client = DeployerClient(settings.deployer_socket_path)
    if not client.available:
        log.warning("deployer socket not available at %s", settings.deployer_socket_path)
        return False

    log.info("submitting init_agent_space_worktree job with WORKSPACE_DIR=%s", repo_path)
    result = client.submit_job(
        {
            "job_id": "init-worktree-bootstrap",
            "job_type": "init_agent_space_worktree",
        },
        args={"WORKSPACE_DIR": repo_path},
    )
    log.info("deployer result: %s", result)
    if result.get("status") == "succeeded":
        log.info("deployer initialized worktree at %s", repo_path)
        return True
    log.warning("deployer init failed: %s", result.get("error", "unknown"))
    return False


def register_system_core_workspace(db: Session) -> Workspace | None:
    """
    Register the agent-space worktree as a system_core workspace in the owner's
    personal space. Idempotent — safe to call on every startup.

    Returns the Workspace record, or None if system evolution is disabled
    or validation fails. Errors are logged, never raised — startup must not
    be blocked by system evolution misconfiguration.
    """
    if not settings.enable_system_evolution:
        return None

    if not settings.system_core_owner_email:
        log.warning(
            "ENABLE_SYSTEM_EVOLUTION=true but SYSTEM_CORE_OWNER_EMAIL is not set — skipping"
        )
        return None

    # Find the owner user by email
    user = db.query(User).filter(User.email == settings.system_core_owner_email).first()
    if not user:
        log.warning(
            "SYSTEM_CORE_OWNER_EMAIL '%s' — no user found with that email; "
            "complete Google OAuth login first — skipping system core registration",
            settings.system_core_owner_email,
        )
        return None

    # Get or create owner's personal space
    space = _ensure_personal_space(db, user)

    # Compute worktree path for this space
    repo_path = paths.system_core_workspace_dir_for_space(space.id).resolve()

    # Ask deployer to clone the canonical repo if not already initialized
    if not _validate_git_repo(repo_path):
        log.info("system_core worktree not found at %s — asking deployer to init", repo_path)
        ok = _call_deployer_init_worktree(str(repo_path))
        if not ok:
            log.warning(
                "system_core workspace '%s' could not be initialized — skipping registration",
                repo_path,
            )
            return None

    if not _validate_git_repo(repo_path):
        log.warning(
            "system_core workspace '%s' is not a valid git repository after init — skipping",
            repo_path,
        )
        return None

    # Check if already registered
    existing = db.query(Workspace).filter(
        Workspace.owner_space_id == space.id,
        Workspace.id == SYSTEM_CORE_WORKSPACE_ID,
    ).first()
    if existing:
        log.info(
            "system_core workspace already registered at %s",
            existing.root_path,
        )
        existing.root_path = str(repo_path)
        db.commit()
        return existing

    # Register the workspace
    ws = Workspace(
        id=SYSTEM_CORE_WORKSPACE_ID,
        owner_space_id=space.id,
        created_by_user_id=user.id,
        name="Agent Space",
        description="Agent-space self-evolution workspace — managed by agent-space",
        workspace_type="system_core",
        kind="repo",
        root_path=str(repo_path),
        visibility="private",
        protected=True,
        system_managed=True,
        registered_from="auto",
        status="active",
    )
    db.add(ws)
    db.commit()
    db.refresh(ws)
    log.info(
        "Registered system_core workspace at %s in space %s",
        repo_path,
        space.id,
    )
    return ws

File: server/src/modules/workspaces/systemCore.ts
/**
 * System-core workspace registration.
 *
 * On server startup (when ENABLE_SYSTEM_EVOLUTION=true), validates that the
 * agent-space repo clone exists, then registers it as a system_core workspace
 * in the owner's personal space. Idempotent — safe to call on every startup.
 */

import { execFile } from "node:child_process";
import { stat } from "node:fs/promises";
import { promisify } from "node:util";
import { join, resolve } from "node:path";
import { randomUUID } from "node:crypto";
import type { ServerConfig } from "../../config";
import { getDbPool, type Pool } from "../../db/pool";
import { DeployerSocketClient } from "../deployment/client";

const execFileAsync = promisify(execFile);

export const SYSTEM_CORE_WORKSPACE_ID = "system-core-workspace";

interface SystemCoreDb {
  query<Row = Record<string, unknown>>(
    sql: string,
    params?: readonly unknown[],
  ): Promise<{ rows: Row[] }>;
}

interface Logger {
  info: (msg: string) => void;
  warn: (msg: string) => void;
}

async function isGitRepo(path: string): Promise<boolean> {
  try {
    await execFileAsync("git", ["rev-parse", "--git-dir"], { cwd: path, timeout: 10_000 });
    return true;
  } catch {
    const gitDir = await stat(join(path, ".git")).catch(() => null);
    return Boolean(gitDir?.isDirectory() || gitDir?.isFile());
  }
}

async function findPersonalSpaceId(pool: Pool, userId: string): Promise<string | null> {
  const res = await pool.query<{ space_id: string }>(
    `SELECT m.space_id
       FROM space_memberships m
       JOIN spaces s ON s.id = m.space_id
      WHERE m.user_id = $1 AND m.status = 'active' AND s.type = 'personal'
      ORDER BY m.created_at ASC, m.id ASC
      LIMIT 1`,
    [userId],
  );
  return res.rows[0]?.space_id ?? null;
}

async function ensurePersonalSpace(pool: Pool, userId: string, displayName: string): Promise<string> {
  const existing = await findPersonalSpaceId(pool, userId);
  if (existing) return existing;
  const spaceId = randomUUID();
  await pool.query(
    `INSERT INTO spaces (id, name, type, created_by_user_id, created_at, updated_at)
     VALUES ($1, $2, 'personal', $3, now(), now())
     ON CONFLICT DO NOTHING`,
    [spaceId, `${displayName}'s Personal Space`, userId],
  );
  await pool.query(
    `INSERT INTO space_memberships (id, space_id, user_id, role, status, created_at, updated_at)
     VALUES ($1, $2, $3, 'owner', 'active', now(), now())
     ON CONFLICT DO NOTHING`,
    [randomUUID(), spaceId, userId],
  );
  return spaceId;
}

export async function upsertSystemCoreWorkspace(
  db: SystemCoreDb,
  input: {
    spaceId: string;
    userId: string;
    workspaceDir: string;
    baseBranch: string;
  },
): Promise<"inserted" | "updated"> {
  const existing = await db.query<{ id: string }>(
    `SELECT id FROM workspaces WHERE id = $1 AND space_id = $2 LIMIT 1`,
    [SYSTEM_CORE_WORKSPACE_ID, input.spaceId],
  );

  if (existing.rows[0]) {
    await db.query(
      `UPDATE workspaces
          SET root_path = $1,
              default_branch = $2,
              workspace_type = 'system_core',
              kind = 'repo',
              status = 'active',
              protected = true,
              system_managed = true,
              registered_from = COALESCE(registered_from, 'auto'),
              created_by_user_id = COALESCE(created_by_user_id, $3),
              owner_user_id = COALESCE(owner_user_id, $3),
              updated_at = now()
        WHERE id = $4 AND space_id = $5`,
      [
        input.workspaceDir,
        input.baseBranch,
        input.userId,
        SYSTEM_CORE_WORKSPACE_ID,
        input.spaceId,
      ],
    );
    return "updated";
  }

  await db.query(
    `INSERT INTO workspaces (
       id, space_id, created_by_user_id, owner_user_id, name, description,
       workspace_type, kind, root_path, default_branch,
       status, visibility, protected, system_managed, registered_from,
       created_at, updated_at
     ) VALUES (
       $1, $2, $3, $3, $4, $5,
       'system_core', 'repo', $6, $7,
       'active', 'private', true, true, 'auto',
       now(), now()
     )`,
    [
      SYSTEM_CORE_WORKSPACE_ID,
      input.spaceId,
      input.userId,
      "agent-space",
      "System core: agent-space self-evolution workspace",
      input.workspaceDir,
      input.baseBranch,
    ],
  );
  return "inserted";
}

/**
 * Register the agent-space repo clone as a system_core workspace in the owner's
 * personal space. Errors are logged, never thrown — startup must not fail due to
 * system evolution misconfiguration.
 */
export async function registerSystemCoreWorkspace(
  config: ServerConfig,
  log: Logger,
): Promise<void> {
  if (!config.enableSystemEvolution) return;
  if (!config.databaseUrl) {
    log.warn("[system_core] no database URL configured — skipping");
    return;
  }

  const ownerEmail = config.systemCoreOwnerEmail ?? config.instanceAdminEmail;
  if (!ownerEmail) {
    log.warn(
      "[system_core] ENABLE_SYSTEM_EVOLUTION=true but neither SYSTEM_CORE_OWNER_EMAIL nor INSTANCE_ADMIN_EMAIL is set — skipping",
    );
    return;
  }

  try {
    const pool = getDbPool(config.databaseUrl);

    const userRes = await pool.query<{ id: string; display_name: string }>(
      `SELECT id, display_name FROM users WHERE LOWER(email) = $1 LIMIT 1`,
      [ownerEmail],
    );
    const user = userRes.rows[0];
    if (!user) {
      log.warn(
        `[system_core] owner email '${ownerEmail}' not found — complete login first, then restart`,
      );
      return;
    }

    const spaceId = await ensurePersonalSpace(pool, user.id, user.display_name);
    const workspaceDir = resolve(config.workspaceRoot, spaceId, "agent-space");

    if (!(await isGitRepo(workspaceDir))) {
      log.info(`[system_core] repo not found at ${workspaceDir} — requesting deployer init`);
      const deployer = new DeployerSocketClient(config);
      const result = await deployer.submit("init_agent_space_worktree", {
        WORKSPACE_DIR: workspaceDir,
      });
      if (result.status !== "succeeded") {
        log.warn(`[system_core] deployer init failed: ${result.error ?? result.status} — skipping`);
        return;
      }
    }

    if (!(await isGitRepo(workspaceDir))) {
      log.warn(`[system_core] ${workspaceDir} is still not a valid git repo after init — skipping`);
      return;
    }

    const action = await upsertSystemCoreWorkspace(pool, {
      spaceId,
      userId: user.id,
      workspaceDir,
      baseBranch: config.systemCoreBaseBranch,
    });
    log.info(
      action === "inserted"
        ? `[system_core] registered system_core workspace at ${workspaceDir} in space ${spaceId}`
        : `[system_core] workspace already registered — refreshed active root_path to ${workspaceDir}`,
    );
  } catch (err) {
    log.warn(
      `[system_core] registration failed: ${err instanceof Error ? err.message : String(err)}`,
    );
  }
}
